Add no-GIL interpreter support

clin1234 opened the pull request.

clin1234 committed:
Add no-GIL interpreter support

Add `pytest-run-parallel` as dependency, test no-GIL interpreters in CI, and
mark Cython module as safe for freethreaded interpreters.

clin1234 committed:
Exclude PYTHON_GIL=0 for now

clin1234 committed:
Lint

clin1234 committed:
Exclude PYTHON_GIL for now

lysnikolaou commented on .github/workflows/wheel.yml:
```diff
@@ -34,6 +34,7 @@ jobs:
           CIBW_TEST_REQUIRES: "pytest"
           CIBW_TEST_COMMAND: "pytest {package}/test"
           CIBW_SKIP: "pp* cp38-macosx_*"
+          CIBW_ENABLE: cpython-freerelease
```
```suggestion
          CIBW_ENABLE: cpython-freethreading
```

lysnikolaou commented on test/test_multithreading.py:
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python3
```
According to https://github.com/msgpack/msgpack-python/issues/613#issuecomment-2914866092, sharing `Packer` objects wouldn't be supported under free-threading, so this test is probably not needed. Could we run the test suite with `pytest-run-parallel` instead?

clin1234 replied:
Whatever works: I borrowed this test from another fork, and I'm still learning how to modernize existing `pytest` tests to use `pytest-run-parallel`

lysnikolaou replied:
What I would do is install `pytest-run-parallel` in CI and then add a step in the main test job that looks something like this:

```yml
- name: Install pytest-run-parallel under free-threading
  if: matrix.py == '3.14t-dev' || matrix.py == '3.13t'
  shell: bash
  run: |
    pip install pytest-run-parallel

- name: Run tests in parallel under free-threading
  if: matrix.py == '3.14t-dev' || matrix.py == '3.13t'
  shell: bash
  run: |
    pytest -v --parallel-threads=auto --iterations=20 test
```

lysnikolaou replied:
Great! I think this file can go now, since we're testing with `pytest-run-parallel` and sharing `Packer` objects is not supported.

lysnikolaou reviewed (commented):
Hey @clin1234! Thanks for working on this. I'm not the maintainer here, but I'm leaving a couple of comments to help move this forward.

clin1234 committed:
Update .github/workflows/wheel.yml

Co-authored-by: Lysandros Nikolaou <lisandrosnik@gmail.com>

clin1234 committed:
Merge branch 'main' into nogil

clin1234 committed:
Fix extremely silly typo

lysnikolaou reviewed (commented):
Thanks for iterating on this so quickly @clin1234! Left some more comments on here. Hope these are helpful!

clin1234 commented on test/test_buffer.py:
```diff
@@ -17,7 +17,7 @@ def test_unpack_bytearray():
     assert [b"foo", b"bar"] == obj
     expected_type = bytes
-    assert all(type(s) == expected_type for s in obj)
+    assert all(type(s) is expected_type for s in obj)
```
Fixed lints as well in this PR

clin1234 committed:
Drop 3.8 support in test.yml

clin1234 committed:
Drop pytest-run-parallel as test requirement in requirements.txt

clin1234 committed:
Separate C extension and fallback tests for GIL and no-GIL interpreters

clin1234 committed:
Work around Windows wonkiness

On Windows, attempting to upgrade `pip` within CI always yields this:
```
Run pip install -U pip
Requirement already satisfied: pip in c:\hostedtoolcache\windows\python\3.10.11\x64\lib\site-packages (25.1.1)
Collecting pip
  Downloading pip-25.2-py3-none-any.whl.metadata (4.7 kB)
Downloading pip-25.2-py3-none-any.whl (1.8 MB)
   ---------------------------------------- 1.8/1.8 MB 48.1 MB/s eta 0:00:00

Notice:  A new release of pip is available: 25.1.1 -> 25.2
Notice:  To update, run: python.exe -m pip install --upgrade pip
ERROR: To modify pip, please run the following command:
C:\hostedtoolcache\windows\Python\3.10.11\x64\python.exe -m pip install -U pip
Error: Process completed with exit code 1.
```

clin1234 committed:
Fix minor YAML oops in .github/workflows/test.yml

Co-authored-by: Frank Dana <ferdnyc@gmail.com>

clin1234 committed:
Fix minor YAML oops in .github/workflows/test.yml

Co-authored-by: Frank Dana <ferdnyc@gmail.com>

clin1234 committed:
Drop spurious multhreading test

Per https://github.com/msgpack/msgpack-python/pull/641#discussion_r2315489310, sharing more than one `Packer` object is not supported.

khneal commented on .github/workflows/test.yml:
```diff
@@ -10,7 +10,7 @@ jobs:
       matrix:
         os: ["ubuntu-latest", "windows-latest", "macos-latest"]
-        py: ["3.14-dev", "3.13", "3.12", "3.11", "3.10", "3.9"]
+        py: ["3.14-dev", "3.14t-dev", "3.13", "3.13t", "3.12", "3.11", "3.10", "3.9"]
```
If "3.13t" takes ANY additional effort, consider leaving it behind and focusing on "3.14t".
[After all](https://docs.python.org/3.14/howto/free-threading-python.html)... "free-threaded mode is experimental" in 3.13

clin1234 replied:
?

ferdnyc replied:
I think @khneal was suggesting to just focus on only 3.14t instead of 3.13t, if supporting both involves more work — but I don't get the impression it does. The same configuration applies to both, and the CI system is the only thing that has to care about whether there's one freethreading test version or two.

clin1234 commented on .github/workflows/test.yml:
```diff
@@ -10,7 +10,7 @@ jobs:
       matrix:
         os: ["ubuntu-latest", "windows-latest", "macos-latest"]
-        py: ["3.14-dev", "3.13", "3.12", "3.11", "3.10", "3.9"]
+        py: ["3.14-dev", "3.14t-dev", "3.13", "3.13t", "3.12", "3.11", "3.10", "3.9"]
```
Will do.

clin1234 committed:
Drop -dev suffix in test.yml

clin1234 committed:
Merge branch 'main' into nogil

methane committed:
Merge branch 'main' into nogil

methane committed:
Merge branch 'main' into nogil
